cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java

```diff
@@ -101,6 +110,60 @@
   }
 
   /**
+   * Migrate schedules in a given namespace from app metadata store to ProgramScheduleStoreDataset
+   *
+   * @param namespaceId  the namespace with schedules to be migrated
+   * @param appMetaStore app metadata store with schedules to be migrated
+   * @return the lexicographically largest namespace id String with schedule migration completed
+   */
+  public String migrateFromAppMetadataStore(NamespaceId namespaceId, Store appMetaStore) {
+    String completedNamespace = getMigrationCompleteNamespace();
+    if (completedNamespace != null && completedNamespace.compareTo(namespaceId.toString()) > 0) {
+      return completedNamespace;
+    }
+    for (ApplicationSpecification appSpec : appMetaStore.getAllApplications(namespaceId)) {
+      ApplicationId appId = namespaceId.app(appSpec.getName(), appSpec.getAppVersion());
+      for (ScheduleSpecification scheduleSpec : appSpec.getSchedules().values()) {
+        String scheduleKey = rowKeyForSchedule(appId, scheduleSpec.getSchedule().getName());
+        ProgramSchedule schedule = Schedulers.toProgramSchedule(appId, scheduleSpec);
+        persistSchedule(scheduleKey, schedule);
+      }
+    }
+    store.put(MIGRATION_COMPLETE_NAMESPACE_ROW_BYTES, MIGRATION_COMPLETE_NAMESPACE_COLUMN_BYTES,
+              Bytes.toBytes(namespaceId.toString()));
+    return namespaceId.toString();
+  }
+
+  @Nullable
+  private String getMigrationCompleteNamespace() {
+    Row row = store.get(MIGRATION_COMPLETE_NAMESPACE_ROW_BYTES);
+    if (row.isEmpty()) {
+      return null;
+    }
+    String namespace = Bytes.toString(row.get(MIGRATION_COMPLETE_NAMESPACE_COLUMN_BYTES));
+    NamespaceId.fromString(namespace); // verify whether the namespace is of correct format
+    return namespace;
+  }
+
+  public void setMigrationComplete() {
+    store.put(MIGRATION_COMPLETE_ROW_BYTES, MIGRATION_COMPLETE_COLUMN_BYTES, Bytes.toBytes(true));
+  }
+
+  public boolean isMigrationComplete() {
+    Row row = store.get(MIGRATION_COMPLETE_ROW_BYTES);
+    if (row.isEmpty()) {
+      return false;
+    }
+    try {
+      return Bytes.toBoolean(row.get(MIGRATION_COMPLETE_COLUMN_BYTES));
+    } catch (Exception e) {
```

Comment:
I don't think we need to handle this exception.
It can only happen if there's data corruption, right? In which case, the exception should be propagated to fail the upgrade, in my opinion, since it is likely indicative of a more severe issue.